### PR TITLE
[gas] initial gas schedule (experimental)

### PR DESCRIPTION
**Pull request by vgao1996**
This is an attempt to come up with some more reasonable values for the gas schedule. 

Highlights:
1. This covers the gas parameters for instructions, transactions and move-stdlib, but it's still missing the ones for the aptos framework
2. A p2p transfer transaction costs 684 gas units. Assuming that the minimum fee per gas unit allowed is 100, then the minimum fee for a p2p transfer transaction is 0.00068456 Aptos.

- - -
This change is [Reviewable](https://reviewable.io/reviews/aptos-labs/aptos-core/4231)

### CODE REVIEW

File: aptos-move/aptos-gas/src/instr.rs [hunk 1]
```diff
@@ -15,176 +15,180 @@ crate::params::define_gas_parameters!(
...
+        [exists_generic_base: InternalGas, "exists_generic.base", 500],
+        [move_from_base: InternalGas, "move_from.base", 500],
         [
             move_from_generic_base: InternalGas,
             "move_from_generic.base",
-            10
+            500
         ],
-        [move_to_base: InternalGas, "move_to.base", 10],
+        [move_to_base: InternalGas, "move_to.base", 500],
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/instr.rs:168]: Does this have to be this expensive? We already charge for storage usage separately right?

File: aptos-move/aptos-gas/src/instr.rs [hunk 2]
```diff
@@ -15,176 +15,180 @@ crate::params::define_gas_parameters!(
...
             "mut_borrow_global_generic.base",
-            10
+            500
         ],
-        [exists_base: InternalGas, "exists.base", 10],
-        [exists_generic_base: InternalGas, "exists_generic.base", 10],
-        [move_from_base: InternalGas, "move_from.base", 10],
+        [exists_base: InternalGas, "exists.base", 500],
+        [exists_generic_base: InternalGas, "exists_generic.base", 500],
+        [move_from_base: InternalGas, "move_from.base", 500],
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/instr.rs:162]: Can this be cheaper than move_to? move_from can be used when deleting resources, which should be cheap if not free (or even give a bonus)

File: aptos-move/aptos-gas/src/instr.rs [hunk 3]
```diff
@@ -15,176 +15,180 @@ crate::params::define_gas_parameters!(
...
         [
             mut_borrow_global_generic_base: InternalGas,
             "mut_borrow_global_generic.base",
-            10
+            500
         ],
-        [exists_base: InternalGas, "exists.base", 10],
-        [exists_generic_base: InternalGas, "exists_generic.base", 10],
-        [move_from_base: InternalGas, "move_from.base", 10],
+        [exists_base: InternalGas, "exists.base", 500],
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/instr.rs:160]: Can this be cheaper? This would penalize assertions.

File: aptos-move/aptos-gas/src/move_stdlib.rs [hunk 4]
```diff
@@ -4,24 +4,26 @@
 use move_stdlib::natives::GasParameters;
 
 crate::natives::define_gas_parameters_for_natives!(GasParameters, "move_stdlib", [
-    [.bcs.to_bytes.per_byte_serialized, "bcs.to_bytes.per_byte_serialized", 1],
-    [.bcs.to_bytes.failure, "bcs.to_bytes.failure", 10],
+    [.bcs.to_bytes.per_byte_serialized, "bcs.to_bytes.per_byte_serialized", 10],
+    [.bcs.to_bytes.failure, "bcs.to_bytes.failure", 1000],
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/move_stdlib.rs:8]: Why is this so expensive?

File: aptos-move/aptos-gas/src/move_stdlib.rs [hunk 5]
```diff
@@ -4,24 +4,26 @@
...
 
-    [.hash.sha2_256.base, "hash.sha2_256.base", 1],
-    [.hash.sha2_256.per_byte, "hash.sha2_256.per_byte", 1],
-    [.hash.sha3_256.base, "hash.sha3_256.base", 1],
-    [.hash.sha3_256.per_byte, "hash.sha3_256.per_byte", 1],
+    [.hash.sha2_256.base, "hash.sha2_256.base", 3000],
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/move_stdlib.rs:10]: Same here

File: aptos-move/aptos-gas/src/instr.rs (hunk 3 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/instr.rs:160]: > Can this be cheaper?
Yes, we can give discounts within a reasonable range in general.

Although I need to point out that the costs here are just for performing the check in memory -- it does not account for the actual storage loads.

File: aptos-move/aptos-gas/src/instr.rs (hunk 2 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/instr.rs:162]: Same as above, we can give a discount.

File: aptos-move/aptos-gas/src/move_stdlib.rs (hunk 5 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/move_stdlib.rs:10]: This is based on benchmarks. There seems to be some "startup costs" associated with these sha functions. Maybe @alinush could explain why?

File: aptos-move/aptos-gas/src/move_stdlib.rs (hunk 4 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/move_stdlib.rs:8]: Ah this is a good one. I guess I did it for no obvious reasons. 

The tricky thing is that bcs can fail, but the failure conditions are hardcoded in the bcs crate itself. :(
The failures only happens in the most extreme cases, like vector having over 2^32 values or values being nested too deep.
I honestly don't think we have a good story for this at all, be it for the bcs native or other parts of the Rust code that uses bcs...

File: aptos-move/aptos-gas/src/transaction.rs [hunk 6]
```diff
@@ -65,36 +65,36 @@ crate::params::define_gas_parameters!(
...
+            100
         ],
         // Gas Parameters for reading data from storage.
-        [load_data_base: InternalGas, "load_data.base", 1],
+        [load_data_base: InternalGas, "load_data.base", 16_000],
         [
             load_data_per_byte: InternalGasPerByte,
             "load_data.per_byte",
-            1
+            1_000
```
Comment (vgao1996) [inline aptos-move/aptos-gas/src/transaction.rs:75]: @movekevin this is where we actually charge for loading resources from storage. This is assessed per resource accessed per transaction.

File: aptos-move/aptos-gas/src/instr.rs (hunk 3 above)
Comment (movekevin) [reply aptos-move/aptos-gas/src/instr.rs:160]: In that case I think move_from, move_to, exists can all be relatively cheap. It shouldn't be much more than other operations.

File: aptos-move/aptos-gas/src/move_stdlib.rs (hunk 4 above)
Comment (movekevin) [reply aptos-move/aptos-gas/src/move_stdlib.rs:8]: I think from_bytes/to_bytes are relatively common operations and should be cheaper if possible. Can we make them be closer to the cost of serializing/deserializing underneath

File: aptos-move/aptos-gas/src/instr.rs (hunk 3 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/instr.rs:160]: To be fair, 500 is not actually a lot of gas, especially considering that we're doing a hash table lookup. Recall that the add instruction currently costs 160, so a global operation is just 3x of that.

I'd be totally fine cutting this down to something like 250 though.

File: aptos-move/aptos-gas/src/transaction.rs (hunk 6 above)
Comment (movekevin) [reply aptos-move/aptos-gas/src/transaction.rs:75]: How did you come up with 1000 gas units per byte?

File: aptos-move/aptos-gas/src/instr.rs (hunk 3 above)
Comment (movekevin) [reply aptos-move/aptos-gas/src/instr.rs:160]: I see that makes sense then

File: aptos-move/aptos-gas/src/move_stdlib.rs (hunk 4 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/move_stdlib.rs:8]: Just to be clear, the "1000" here is the failure cost and as I mentioned above, a failure should be extremely rare...

Or is it that you think the unit cost is too high?

**Review by wrwg — COMMENTED** [review]
What is blocking us to move forward with this?

Maybe for the aptos_framework natives, we can fill in some dummies to make it at least relative more consistent. For example all the hashing functions should cost as much as sha256 or something.

**Comment by movekevin** [comment]
> What is blocking us to move forward with this?
> 
> Maybe for the aptos_framework natives, we can fill in some dummies to make it at least relative more consistent. For example all the hashing functions should cost as much as sha256 or something.

I'm okay with moving forward with this since we're going to keep iterating on things. Only thing here is to make sure the entire ecosystem still works after these changes

**Review by movekevin — APPROVED** [review]

File: aptos-move/aptos-gas/src/transaction.rs (hunk 6 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/transaction.rs:75]: I borrowed the logic/ratio from EVM. In EVM, loading a word (32B) costs 100 gas units, while add costs 3. In our current implementation, the simpler instructions cost anywhere around 60 to 200, so I thought 1000 per byte might be reasonable.

I'm open to change it if you feel like's not right, and my hope is that we'll be able to fine tune it as we gather more data.

**Review by wrwg — APPROVED** [review]

File: aptos-move/aptos-gas/src/transaction.rs (hunk 6 above)
Comment (movekevin) [reply aptos-move/aptos-gas/src/transaction.rs:75]: Okay makes sense

File: aptos-move/aptos-gas/src/move_stdlib.rs (hunk 4 above)
Comment (movekevin) [reply aptos-move/aptos-gas/src/move_stdlib.rs:8]: Yeah I think it might be a bit high. But we can tune it later

Comment (davidiw) [reply aptos-move/aptos-gas/src/move_stdlib.rs:8]: failure is impossible in bcs, imo

File: aptos-move/aptos-gas/src/transaction.rs [hunk 7]
```diff
@@ -20,7 +20,7 @@ crate::params::define_gas_parameters!(
         [
             min_transaction_gas_units: InternalGas,
             "min_transaction_gas_units",
-            600
+            1_500_000
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/transaction.rs:23]: What is this for?

File: aptos-move/aptos-gas/src/transaction.rs [hunk 8]
```diff
@@ -33,7 +33,7 @@ crate::params::define_gas_parameters!(
         [
             intrinsic_gas_per_byte: InternalGasPerByte,
             "intrinsic_gas_per_byte",
-            8
+            2_000
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/transaction.rs:36]: Why is this changed to 2000 now?

File: aptos-move/aptos-gas/src/transaction.rs [hunk 9]
```diff
@@ -65,36 +65,36 @@ crate::params::define_gas_parameters!(
         [
             gas_unit_scaling_factor: GasScalingFactor,
             "gas_unit_scaling_factor",
-            1000
+            10_000
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/transaction.rs:68]: Why this change?

File: aptos-move/aptos-gas/src/transaction.rs [hunk 10]
```diff
@@ -65,36 +65,36 @@ crate::params::define_gas_parameters!(
...
         ],
         // Gas Parameters for reading data from storage.
-        [load_data_base: InternalGas, "load_data.base", 1],
+        [load_data_base: InternalGas, "load_data.base", 16_000],
```
Comment (movekevin) [inline aptos-move/aptos-gas/src/transaction.rs:71]: Why 16k?

**Review by movekevin — APPROVED** [review]
No blocking comments as we'll be iterating more on this in the next 1-2 weeks. I just want to start the discussions on a couple of values I see.

File: aptos-move/aptos-gas/src/transaction.rs (hunk 9 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/transaction.rs:68]: After a discussion with Avery, I realized that it's important we preserve some headroom for future appreciation of the coin. With the scaling factor being set to 10000, a p2p transfer transaction costs ~700, which has enough precision.

File: aptos-move/aptos-gas/src/transaction.rs (hunk 8 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/transaction.rs:36]: This is to account for things like deserialization and bytecode verification

File: aptos-move/aptos-gas/src/transaction.rs (hunk 7 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/transaction.rs:23]: This is the minimum gas fee any transaction would need to pay, even if they fail early in the pipeline.

File: aptos-move/aptos-gas/src/transaction.rs (hunk 10 above)
Comment (vgao1996) [reply aptos-move/aptos-gas/src/transaction.rs:71]: Guesswork. I just multiplied the per-byte cost by a reasonable factor. Feel free to suggest alternatives
